### PR TITLE
Separate out a Daydream backend

### PR DESCRIPTION
**Pull request by liaxim**
+ some viewManager refactor

GearVRf-DCO-1.0-Signed-off-by: Mihail Marinov <m.marinov@samsung.com>

### CODE REVIEW

**Comment by liaxim** [comment]
Basic work done; lots of testing left.

We will have three backends; order of loading is determined by the asset/backend_x.txt file where x is the priority. This file gives a factory class name we use to bring up the backend and hook it up.

The monoscopic renderers are still bundled.

Significant refactoring of viewManager but there is more left when it comes to dealing with screenshots, statistics.

**Comment by liaxim** [comment]
Smoke testing showed only a problem with shadow mapping.

**Comment by liaxim** [comment]
The problem with gvr-shadows was caused by another change. Fix merged in master already.

**Comment by liaxim** [comment]
gvr-lodtest doesn't work; doesn't work prior to this change too. Looks to be caused by the recent bounding volume change. Will look into it anyway.

**Comment by liaxim** [comment]
All samples works as well as before. I will experiment with the existing unit tests but this is ready for review already.

lodtest is a separate thing that will have to covered by another pull request.

**Comment by rahul27** [comment]
Will take a look.

File: GVRf/Framework/backend_oculus/src/main/java/org/gearvrf/OvrSurfaceView.java [hunk 1]
```diff
@@ -93,12 +93,6 @@ public OvrSurfaceView(Context context,
     }
 
     @Override
-    public boolean onKeyDown(int keyCode, KeyEvent event) {
```
Comment (rahul27) [inline GVRf/Framework/backend_oculus/src/main/java/org/gearvrf/OvrSurfaceView.java]: Was this intentionally removed?

File: GVRf/Framework/backend_oculus/build.gradle [hunk 2]
```diff
@@ -136,8 +132,6 @@ android {
 dependencies {
     compile fileTree(include: ['*.jar'], dir: 'libs')
     compile fileTree(include: ['*.jar'], dir: 'src/main/libs')
-    compile(name: 'google-vr-base', ext: 'aar')
-    compile(name: 'google-vr-common', ext: 'aar')
     compile project(':framework')
 }
```
Comment (rahul27) [inline GVRf/Framework/backend_oculus/build.gradle]: Cant comment on the next file, will add it here instead:

Why is the oculus backend being loaded at order 2?

File: GVRf/Framework/framework/src/main/AndroidManifest.xml [hunk 3]
```diff
@@ -31,8 +31,6 @@
         android:minSdkVersion="19"
         android:targetSdkVersion="19" />
 
-    <uses-permission android:name="com.samsung.android.hmt.permission.READ_SETTINGS" />
```
Comment (rahul27) [inline GVRf/Framework/framework/src/main/AndroidManifest.xml]: Has this been moved to the oculus backend. Don't see it there.

**Comment by liaxim** [comment]
I got caught by .gitignore. Will add the daydream backend files.

File: GVRf/Framework/framework/src/main/AndroidManifest.xml (hunk 3 above)
Comment (liaxim) [inline GVRf/Framework/framework/src/main/AndroidManifest.xml]: It was duplicated in this file itself.

File: GVRf/Framework/backend_oculus/build.gradle (hunk 2 above)
Comment (liaxim) [inline GVRf/Framework/backend_oculus/build.gradle]: Leaving room for another backend to have higher loading priority.

File: GVRf/Framework/backend_oculus/src/main/java/org/gearvrf/OvrSurfaceView.java (hunk 1 above)
Comment (liaxim) [inline GVRf/Framework/backend_oculus/src/main/java/org/gearvrf/OvrSurfaceView.java]: Yes. It is going to the daydream backend module.

**Comment by liaxim** [comment]
Ran framework-tests; I don't think they are yet in a state that allows conclusive decisions.

File: GVRf/Framework/backend_daydream/src/main/java/org/gearvrf/DaydreamViewManager.java [hunk 4]
```diff
@@ -28,52 +28,49 @@
...
+        updateSensoredScene();
     }
 
     public void onDrawEye(Eye eye) {
         if (eye.getType() == Eye.Type.LEFT) {
-            OvrMonoscopicRenderer.cull(mMainScene, mMainScene.getMainCameraRig().getCenterCamera
-                    (), mRenderBundle);
-            renderCamera(mActivity.getNative(), mMainScene, mMainScene
-                    .getMainCameraRig().getLeftCamera(), mRenderBundle);
+//            OvrMonoscopicRenderer.cull(mMainScene, mMainScene.getMainCameraRig().getCenterCamera
```
Comment (liaxim) [inline GVRf/Framework/backend_daydream/src/main/java/org/gearvrf/DaydreamViewManager.java]: Removed commented out code from the daydream viewManager. Culling (and shadow mapping) is done by the base viewManager for all backends.

**Comment by liaxim** [comment]
Monoscopic and Nexus is something we can't handle right now. It is a rather special case since Oculus is not there but the monoscopic renderer is bundled with Oculus.

This needs work of separating out the monoscopic renderer(s). @thomasflynn Should this be handled as part of this change or could it be done in the future?

**Comment by thomasflynn** [comment]
future

**Comment by liaxim** [comment]
Rahul is ok with merging this. Will merge tomorrow morning if there are no further comments.

**Comment by NolaDonato** [comment]
I tested this today with the experimental gradle build and the normal build. It works OK for the Oculus backend. I don't have a Daydream phone so I did not test that option.

**Pull request merged**